athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/AwsRestHighLevelClientFactory.java

```diff
@@ -33,38 +35,71 @@
 {
     private static final Logger logger = LoggerFactory.getLogger(AwsRestHighLevelClientFactory.class);
 
-    private final boolean useAwsCredentials;
-
     /**
      * credentialsPattern is used for parsing out the username/password credentials out of the endpoint with the
      * format: "username@password:" (e.g. if the endpoint is http://myusername@mypassword:www.google.com, then this
      * pattern will be used to extract the following credentials string: "myusername@mypassword:").
      */
     private static final Pattern credentialsPattern = Pattern.compile("[^/@]+@[^:]+:");
 
+    /**
+     * A flag indicating how the created clients will authenticate to an Elasticsearch instance.
+     * If useAwsCredentials = true, the client will be injected with AWS credentials.
+     * If useAwsCredentials = false and username/password are extracted using the credentialsPattern, the client
+     * will be injected with username/password credentials.
+     * If useAwsCredentials = false and username/password are not present, the client will be created with no
+     * credentials.
+     */
+    private final boolean useAwsCredentials;
+
+    /**
+     * Stored cache of clients accessible via the endpoint.
+     */
+    private final Map<String, AwsRestHighLevelClient> clientCache;
+
     /**
      * Constructs a new client factory (using a builder) that will create clients injected with credentials.
      * @param useAwsCredentials if true the factory create clients injected with AWC credentials.
-     *                              If false, it will create clients injected with username/password credentials.
+     *                          If false, it will create clients injected with username/password credentials.
      */
-    protected AwsRestHighLevelClientFactory(boolean useAwsCredentials)
+    public AwsRestHighLevelClientFactory(boolean useAwsCredentials)
     {
         this.useAwsCredentials = useAwsCredentials;
+        this.clientCache = new HashMap<>();
```

Comment:
We shouldn't' allow clients to be unbounded size.

Reply:
CacheableAwsRestHighLevelClient has a max size of 10 clients.